core/src/main/java/org/apache/calcite/sql/fun/SqlCoalesceFunction.java

```diff
@@ -80,4 +93,40 @@ public SqlCoalesceFunction() {
     assert call.getFunctionQuantifier() == null;
     return SqlCase.createSwitched(pos, null, whenList, thenList, elseExpr);
   }
+
+  @Override @NonNull public RelDataType inferReturnType(
```

Comment:
I am afraid that this solution has the same problems as your previous one: it modifies (using type coercion) operand types during the *return* type inference. I think you are only allowed to change the operand types during *operand* type inference.

Reply:
strange, I see that Case is in fact doing the same thing

Reply:
I take back my comment above and I will approve this solution.
I guess I don't yet understand completely how the type checker works.
I see that `SqlOperator.validateOperands` calls `checkOperandTypes`, then `inferReturnType`, which in turn may call `operandTypeInference`. This is strange - inferring operand types after they have already been checked.
But we won't fix this problem in this PR.